File: algorithms/code/mingrammer/mergesort.py
def mergesort(arr):
    size = len(arr)

    if size == 1:
        return arr

    l_arr = mergesort(arr[:size // 2])
    r_arr = mergesort(arr[size // 2:])

    l_size = len(l_arr)
    r_size = len(r_arr)

    i, j = 0, 0

    merged = []

    while i < l_size and j < r_size:
        if l_arr[i] < r_arr[j]:
            merged.append(l_arr[i])
            i += 1
        else:
            merged.append(r_arr[j])
            j += 1

    merged.extend(l_arr[i:])
    merged.extend(r_arr[j:])

    return merged

File: algorithms/code/mingrammer/mergesort.go
package sort

func mergeSort(arr []int) []int {
	size := len(arr)

	if size == 1 {
		return arr
	}

	lArr := mergeSort(arr[:size/2])
	rArr := mergeSort(arr[size/2:])

	lSize := len(lArr)
	rSize := len(rArr)

	i, j := 0, 0

	var merged []int

	for i < lSize && j < rSize {
		if lArr[i] <= rArr[j] {
			merged = append(merged, lArr[i])
			i++
		} else {
			merged = append(merged, rArr[j])
			j++
		}
	}

	merged = append(merged, lArr[i:]...)
	merged = append(merged, rArr[j:]...)

	return merged
}
